man/man3/h2devLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 21, 2005
.Dt errorLib 3
.Os pocoLibs
.Sh NAME 
.Nm errorLib
.Nd error code manipulation for pocoLibs
.Sh SYNOPSIS
.Fd "#include <portLib.h>"
.Fd "#include <h2devLib.h>"
.Ft STATUS
.Fn h2devInit void
.Ft int 
.Fn h2devAlloc "char *name" "H2_DEV_TYPE type"
.Ft int 
.Fn h2devFind "char *name" "H2_DEV_TYPE type"
.Ft void 
.Fn h2devShow void
.Ft void 
.Fn h2devFree "int dev"
.Ft STATUS 
.Fn h2devEnd void
.Ft STATUS 
.Fn h2devClean "char *pattern"
.Sh DESCRIPTION
.Fn h2devInit
initializes a 
.SM pocolibs
session.
It creates the IPC structures used by the library and initializes them 
to a know state.
.Pp
.Fn h2devInit 
should be called once, by any process, before using any other
.SM pocoLibs
function. This is generally handled by the 
.Xr h2 1
application.
.Fn h2devAlloc
allocates a
.Nm pocoLib
device.
The 
.Fa name
parameter specifies a name for the device to be allocated, and the 
.Fa type
parameter specifies the type of the device. It can be
.Dv H2_DEV_TYPE_SEM 
to create a semaphore device,
.Dv H2_DEV_TYPE_TASK
to create a task device,
.Dv H2_DEV_TYPE_MBOX
to create a mailbox device or
.Dv H2_DEV_TYPE_POSTER
to create a poster device.
.Pp
.Fn h2devFind
searches for a 
.Nm pocoLib
device with name 
.Fa name  
and type
.Fa type ,
allocated by another process, in order to access it.
.Pp
.Fn h2devShow
outputs to stdout a list of currently allocated 
.SM pocoLibs
devices.
.Pp
.Fn h2devFree
frees a 
.Nm pocoLib
device.
The corresponding device is returned to the system.
.Pp
.Fn h2devEnd
terminates a 
.Nm pocoLibs 
session.
It destroys any existing 
.Nm pocoLib
device and frees all used IPC structures (semaphores, mailboxes or shared 
memory) 
.Pp
All processes using 
.SM pocoLibs
should be terminated before calling
.Fn h2devEnd .
This is generally handled by the 
.Xr h2 1
application.
.Pp
.Fn h2devClean
searches for 
.Nm pocoLib
devices matching
.Fa pattern
(in the sense of shell file globing) and delete them.
.Sh RETURN VALUES
.Fn h2devAlloc
returns the index of the newly allocated device upon success, or
.Dv ERROR
in case of an error. In this case it calls
.Xr errnoSet 3
with the corresponding Unix or pocoLib error code.
.Pp
.Fn h2devFind
returns the index of the found device upon success, or
.Dv ERROR
in case of an error. In this case it calls
.Xr errnoSet 3
with the corresponding Unix or pocoLib error code.
.Sh SEE ALSO
.Xr h2 1 ,
.Xr h2semLib 3 ,
.Xr mboxLib 3 ,
.Xr posterLib 3 .


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/comLib/h2devLib.c
/*
 * Copyright (c) 2004-2005
 *      Autonomous Systems Lab, Swiss Federal Institute of Technology.
 * Copyright (c) 1990, 2003-2005 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "pocolibs-config.h"


#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "portLib.h"

#include <fnmatch.h>

#include "errnoLib.h"
#include "h2semLib.h"
#include "mboxLib.h"
#include "h2errorLib.h"
#include "h2devLib.h"
static const H2_ERROR h2devLibH2errMsgs[] = H2_DEV_LIB_H2_ERR_MSGS;

#include "smMemLib.h"
#include "smObjLib.h"


#ifdef COMLIB_DEBUG_H2DEVLIB
# define LOGDBG(x)	logMsg x
#else
# define LOGDBG(x)
#endif

/* Local fucntions prototypes */
static int h2devFindAux(const char *name, H2_DEV_TYPE type);

/*----------------------------------------------------------------------*/

/**
 ** Record errors messages
 **/
int
h2devRecordH2ErrMsgs(void)
{
    return h2recordErrMsgs("h2devRecordH2ErrMsg", "h2devLib", M_h2devLib, 
			   sizeof(h2devLibH2errMsgs)/sizeof(H2_ERROR), 
			   h2devLibH2errMsgs);
}

/**
 ** Allocation d'un device h2
 **/
int
h2devAlloc(const char *name, H2_DEV_TYPE type)
{
    int i;

    if (h2devAttach() == ERROR) {
	return ERROR;
    }
    h2semTake(0, WAIT_FOREVER);

    /* Verifie que le nom n'existe pas */
    if (type != H2_DEV_TYPE_SEM && h2devFindAux(name, type) != ERROR) {
	h2semGive(0);
	errnoSet(S_h2devLib_DUPLICATE_DEVICE_NAME);
	return ERROR;
    }
    /* Recherche un device libre */
    for (i = 0; i < H2_DEV_MAX; i++) {
	if (h2Devs[i].type == H2_DEV_TYPE_NONE) {
	    /* Trouve' */
	    strncpy(h2Devs[i].name, name, H2_DEV_MAX_NAME);
	    h2Devs[i].type = type;
	    h2Devs[i].uid = getuid();
	    h2semGive(0);
	    LOGDBG(("comLib:h2devAlloc: created device %d\n", i));
	    return i;
	}
    } /* for */

    /* Pas de device libre */
    h2semGive(0);
    errnoSet(S_h2devLib_FULL);
    return ERROR;
}

/*----------------------------------------------------------------------*/

/**
 ** Liberation d'un device h2
 **/
STATUS
h2devFree(int dev)
{
    uid_t uid = getuid();

    if (h2devAttach() == ERROR) {
	return ERROR;
    }
    if (uid != H2DEV_UID(dev) && uid != H2DEV_UID(0)) {
	errnoSet(S_h2devLib_NOT_OWNER);
	return ERROR;
    }
    h2semTake(0, WAIT_FOREVER);
    h2Devs[dev].type = H2_DEV_TYPE_NONE;
    h2semGive(0);
    return OK;
}


/*----------------------------------------------------------------------*/

/**
 ** Clean h2 devices matching glob pattern `name'.
 **/

STATUS
h2devClean(const char *name)
{
   int i, match = 0;
   unsigned char *pool;

   if (h2devAttach() == ERROR) {
      return ERROR;
   }
   /* Look for devices */
   for (i = 0; i < H2_DEV_MAX; i++) {
      if (H2DEV_TYPE(i) != H2_DEV_TYPE_NONE && 
	  fnmatch(name, H2DEV_NAME(i), 0) == 0) {
	 logMsg("Freeing %s\n", H2DEV_NAME(i));
	 match++;
	 switch (H2DEV_TYPE(i)) {
	      case H2_DEV_TYPE_MBOX:
		 mboxDelete(i);
		 break;
	      case H2_DEV_TYPE_POSTER:
		pool = smObjGlobalToLocal(H2DEV_POSTER_POOL(i));
		if (pool != NULL) 
		    smMemFree(pool);
		h2semDelete(H2DEV_POSTER_SEM_ID(i));
		h2devFree(i);
		break;
	      case H2_DEV_TYPE_TASK:
		h2semDelete(H2DEV_TASK_SEM_ID(i));
		h2devFree(i);
		break;
	      case H2_DEV_TYPE_SEM:
	      case H2_DEV_TYPE_NONE:
		break;
	      default:
		/* error */
		logMsg("comLib: unknown device type %d\n", H2DEV_TYPE(i));
		return ERROR;
		break;
	    } /* switch */
	} 
    } /* for */

    if (match == 0) {
	logMsg("No matching device\n");
	return ERROR;
    }
    return OK;
}


/*----------------------------------------------------------------------*/

/**
 ** Recherche d'un device h2
 **/

static int 
h2devFindAux(const char *name, H2_DEV_TYPE type)
{
    int i;

    for (i = 0; i < H2_DEV_MAX; i++) {
	if ((type == h2Devs[i].type) 
	    && (strcmp(name, h2Devs[i].name) == 0)) {
	    return(i);
	}
    } /* for */
    return ERROR;
}

int
h2devFind(const char *name, H2_DEV_TYPE type)
{
    int i;

    if (name == NULL) {
	errnoSet(S_h2devLib_BAD_PARAMETERS);
	return ERROR;
    }
    if (h2devAttach() == ERROR) {
	return ERROR;
    }
    h2semTake(0, H2DEV_TIMEOUT);
    i = h2devFindAux(name, type);
    h2semGive(0);

    if (i >= 0) {
	return i;
    } else {
	errnoSet(S_h2devLib_NOT_FOUND);
	return ERROR;
    }
}

/*----------------------------------------------------------------------*/

/**
 ** Retourne le semId des semaphores
 **/
int
h2devGetSemId(void)
{
    if (h2devAttach() == ERROR) {
	return ERROR;
    }
    if (h2Devs[0].type != H2_DEV_TYPE_SEM) {
	errnoSet(S_h2devLib_BAD_DEVICE_TYPE);
	return ERROR;
    }
    return h2Devs[0].data.sem.semId;
}
